src/utils/renderHelpers.ts

```diff
@@ -0,0 +1,108 @@
+import { Node } from '../ast/nodes/shared/Node';
+import MagicString from 'magic-string';
+import { RenderOptions } from '../Module';
+
+export function findFirstOccurrenceOutsideComment (code: string, searchString: string) {
+	let codeStart = 0;
+	let commentStart, commentLength, lineBreakPos;
+	while (true) {
+		commentStart = code.indexOf('/');
+		lineBreakPos = (~commentStart ? code.slice(0, commentStart) : code).indexOf(searchString);
+		if (~lineBreakPos || !~commentStart) {
+			break;
+		}
+		code = code.slice(commentStart + 1);
+		codeStart += commentStart + 1;
+		if (code[0] === '*') {
+			commentLength = code.indexOf('*/') + 2;
+		} else if (code[0] === '/') {
+			commentLength = code.indexOf('\n') + 1;
+		} else {
+			continue;
+		}
+		code = code.slice(commentLength);
+		codeStart += commentLength;
+	}
+	return ~lineBreakPos ? codeStart + lineBreakPos : -1;
+}
+
+export function findFirstLineBreakOutsideComment (code: string) {
+	let codeStart = 0;
+	let commentStart, commentLength, lineBreakPos;
+	while (true) {
+		commentStart = code.indexOf('/*');
+		lineBreakPos = (~commentStart ? code.slice(0, commentStart) : code).indexOf('\n');
+		if (~lineBreakPos || !~commentStart) {
+			break;
+		}
+		code = code.slice(commentStart);
+		commentLength = code.indexOf('*/') + 2;
+		code = code.slice(commentLength);
+		codeStart += commentStart + commentLength;
+	}
+	return ~lineBreakPos ? codeStart + lineBreakPos : -1;
+}
+
+export function renderStatementList (statements: Node[], code: MagicString, start: number, end: number, options: RenderOptions) {
+	if (statements.length === 0) return;
+	let currentNode, currentNodeStart;
+	let nextNode = statements[0];
+	let nextNodeStart = start + findFirstLineBreakOutsideComment(code.original.slice(start, nextNode.start)) + 1;
+
+	for (let nextIndex = 1; nextIndex <= statements.length; nextIndex++) {
+		currentNode = nextNode;
+		currentNodeStart = nextNodeStart;
+		nextNode = statements[nextIndex];
+		nextNodeStart = currentNode.end + findFirstLineBreakOutsideComment(
+			code.original.slice(currentNode.end, nextNode === undefined ? end : nextNode.start)
+		) + 1;
+		if (currentNode.included) {
+			currentNode.render(code, options, { start: currentNodeStart, end: nextNodeStart });
```

Comment:
I actually would rather leave it here as it is for now because there are actually more parameters I see floating around here in the future. For instance when optimizing sequence expressions, conditionals and logical expressions, we need to pass in if the parent is a call expression to be able to preserve the context. Nevertheless, `isStatement` sounds like a better name for this variable. I would leave this as is for now and try some other performance improvements first.

Reply:
Sure, and difficult to assume what will work without measurement too.